docs/user-guide/reduction-workflow-guidelines.md

```diff
@@ -0,0 +1,139 @@
+
+**Example**
+- If time-of-flight is single-precision, wavelength and momentum transfer should be single-precision.
+- In counts are single-precision, reduced intensity should be single-precision.
+
+### S.5: Switches to double-precision shall be deliberate, explicit, and documented
+
+**Reason**
+Some workflows may require switching to double-precision at a certain point in the workflow.
+This should be a deliberate choice, and the reason for the switch should be documented.
+
+### S.6: Propagation of uncertainties in broadcast operations should support "drop" and "upper-bound" strategies, "upper-bound" shall be the default
+
+**Reason**
+Unless explicitly computed, the exact propagation of uncertainties in broadcast operations is not well-defined.
+Dropping uncertainties is not desireable in general, as it may lead to underestimation of the uncertainties, but we realize that the upper-bound approach may not be suitable in all cases.
+We should therefore support two strategies, "drop" and "upper-bound", and "upper-bound" should be the default.
+
+**Note**
+See [Systematic underestimation of uncertainties by widespread neutron-scattering data-reduction software](http://dx.doi.org/10.3233/JNR-220049) for a discussion of the topic.
+TODO Add reference to upper-bound approach.
+
+### S.7: Do not write files or make write requests to services such as SciCat in providers
+
+**Reason**
+Providers should be side-effect free, and should not write files or make write requests to services.
+
+**Note**
+Workflow may run many times, or in parallel, or task may be retried after failure, and we want to avoid side-effects in these cases.
+This will, e.g., avoid unintentional overwriting of a user's files.
+
+### S.8: Detector banks shall be loaded with their logical dimensions, if possible
+
+**Reason**
+Using logical dims (instead of a flat list of pixels) allows for simpler indexing and slicing of the data, reductions over a subset of dimensions, and masking of physical components.
+
+**Note**
+This is not always possible, as some detectors have an irregular structure and cannot be reshaped to a (multi-dimensional) array.
+
+## T: Testing
+
+### T.1: Adherence to the guidelines shall be tested, and the guideline ID shall be referenced in the test name
+
+**Reason**
+We want to ensure that the guidelines are followed, and that this remains the case as the code base evolves.
+Referencing the guideline ID in the test name makes it easier to find the relevant guideline (or vice-versa), or remove the test if the guideline is removed.
+
+**Note**
+Note all guidelines are testable.
+
+### T.2: Write unit tests for providers
+
+**Reason** Unit tests for providers are easier to write and maintain than for entire workflows.
+
+**Note** This does not mean that we should not write tests for entire workflows, but that we should also write tests for providers.
```

Comment:
How about a guideline for structuring the test directory? And for how to select whether to run integration tests or not. If they take a long time to run, it would be useful to be able to easily deselect them.

Reply:
I think splitting unit tests from integration tests (or something like that) is a great suggestion! As we have not done that before, maybe we can add it in a future version, once we have tried it out in a project?